### PR TITLE
#164700316 As a bot, I should be able to send text messages with any credentials via Twilio

### PR DESCRIPTION
**Pull request by TheoOkafor**
### What does this PR do?
Enable bot to use orgs credentials

### Description of tasks to be completed
- Remove botkit implementation for receiving sms
- Create custom implementation for receiving sms
- Enable bot to fetch org credentials and use it for sending sms
- Add node clusters to enhance bot performance

### How should this be manually tested?
* Clone this repo
* Checkout to a branch `ft-use-orgs-twilio-sms-credentials-164700316`.

* Run `npm start` and `npm run start:live:dev`
* Setup Twilio webhook.
* Log in as a coach and create a client if not already present.
* Add a task to the client's work plan.
* Visit [twilio-gui](http://twilio-gui.herokuapp.com) and get a phone number.
* login to `twilio-gui`
* Using the direct message feature from the coach app, send a message to the client
* Notice that the client did not receive the message.

**Test For When the credential is provided**
* Update your local org account with the available Twilio credentials.
* Using the direct message feature from the coach app, send a message to the client
* Notice that the client receives the message.

**Test For Development Usage**
* Add `NODE_ENV=development` to the local `.env` file.
* Using the direct message feature from the coach app, send a message to the client
* Notice that the client receives the message.

### What are the relevant pivotal tracker stories?
[#164700316](https://www.pivotaltracker.com/story/show/164700316)

### Any Background Context
"Right now, IDEO.org is paying for every text sent to/from Roo. Our plan is to open the app up to the public, and that means a lot more texts...and a lot more money. But we don't have that money!

in the public version, we are asking admins to create their own Twilio account and to enter their Twilio credentials in their org settings (saved to the Org table in the DB).

The bot should use the saved credentials for any text message sent to/from a client of an organization. Here is an example.

> Tracy is a client of ABC Financial Services in Texas. When Roo sends Tracy a text, it uses the Twilio credentials for ABC Financial Services, and they receive their own bill at the end of the month.

> Joseph is a client of XYZ Financial Services in California. When Roo sends Joseph a text, it uses the Twilio credentials for XYZ Financial Services, and they will receive their own bill at the end of the month.
" - @matthewepler 

### Screenshots?
N/A

### Questions
N/A

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -5,19 +5,54 @@ const Messenger = require('./src/Messenger');
 const constants = require('./src/constants');
 require('dotenv').config();
 const api = require('./src/api');
-const Botkit = require('botkit');
 const server = require('./server.js');
 const helpers = require('./helpers');
+const cluster = require('cluster');
```
Comment (TheoOkafor) [inline index.js]: I introduced node clusters to enhance the performance of the app. It optimizes the use of the available CPUs in the machine. It also makes the bot more available by serving multiple instances of the bot in the various workers.

For more information see [How to Create a Node.js Cluster for Speeding Up Your Apps](https://www.sitepoint.com/how-to-create-a-node-js-cluster-for-speeding-up-your-apps/) and [Scaling-out with Node Clusters](https://medium.com/the-andela-way/scaling-out-with-node-clusters-1dca4a39a2a) (An Andela Way article).

File: src/Messenger.js [hunk 2]
```diff
@@ -209,13 +209,40 @@ function sendFBMessage(userId, message, isMessageSentFromCheckIn) {
...
+    TWILIO_AUTH_TOKEN,
+    TWILIO_NUMBER,
+    NODE_ENV,
+  } = process.env;
+  const orgCredentials = await api.getOrg(orgId);
+  let twilioAccountSid = orgCredentials.account_sid;
+  let twilioAuthToken = orgCredentials.auth_token;
+  let twilioPhoneNumber = orgCredentials.twilio_number;
+  
+  if (NODE_ENV === 'development'
```
Comment (65black) [inline src/Messenger.js]: I think a single if check would work here.

```
if (NODE_ENV !== production) {
  // continue
}
```

File: src/api.js [hunk 3]
```diff
@@ -48,6 +48,28 @@ async function getOrgName(id) {
   return null;
 }
 
+/**
+ * Fetches the details of an organization from the steps app database
+ * @param {number} id 
+ */
```
Comment (65black) [inline src/api.js]: Please add a `@returns` tag for this JSDoc seeing as this function returns something.

File: src/api.js [hunk 4]
```diff
@@ -48,6 +48,28 @@ async function getOrgName(id) {
...
+async function getOrg(id) {
+  const org = await rp({
+    method: 'GET',
+    uri: `${process.env.API_URL}/orgs/${id.toString()}`,
+    headers: {
+      Authorization: `Bearer ${process.env.OAUTH_ACCESS_TOKEN}`,
+    },
+    json: true
+  }).catch((e) => {
+    console.log('getOrg api method failed for org id ' + id, e.message);
```
Comment (65black) [inline src/api.js]: I don't think this `console.log` is necessary as we're already handling the error.

Comment (TheoOkafor) [reply src/api.js]: The console.log is actually necessary. We have a Heroku add-on (paper-trial) that caches the console errors and saves them on our dashboard and also forward it to roo email, to enable us to understand what might have gone wrong after the app has been deployed.

File: src/Messenger.js (hunk 2 above)
Comment (TheoOkafor) [reply src/Messenger.js]: You're right @65black. Thank you!

File: src/api.js (hunk 3 above)
Comment (TheoOkafor) [reply src/api.js]: Done

File: index.js [hunk 5]
```diff
@@ -1,23 +1,102 @@
...
+  const userPlatformId = message.From;
+  const userMessage = message.Body;
+  await run({
+    platform: constants.SMS,
+    userPlatformId,
+    userMessage
+  });
+}
+
+async function testTwilioCredentialsController (request, response) {
```
Comment (andela-kakpobome) [inline index.js]: Does this method really need to be in the codebase?

Comment (TheoOkafor) [reply index.js]: Yes.
A better alternative would have been implementing Twilio in the coach app and to use the twilio SMS API there. When I tried it, some dependencies in the Twilio npm package causes the coach app server to fail on `npm start`.

This was the only option I could think of.

**Pull request merged**